### DISCUSSION TITLE
Compiling arm64 ipxe file to boot arm64 devices

### DISCUSSION THREAD

**Discussion by jamaya77** [ipxe/ipxe#1308] [Q&A] [answered]
Hi,

Currently, my ipxe server only supports x86 and x64 devices, but also needed to support arm64. I followed the steps of a previous [discussion](https://github.com/ipxe/ipxe/issues/1088), but its generating a different message on my end of a missing file in the "config/local/" directory. The file does exist, however, in the "config" directory.

/ipxe/src/config/local

[root@ipxe src]# make CROSS_COMPILE=aarch64-linux-gnu- ARCH=arm64 bin-arm64-efi/ipxe.efi EMBED=boot.ipxe -C /ipxe/src && cp ipxe/src/bin-x86_64-efi/ipxe.efi /tftpboot/arm_uefi/
make: Entering directory '/ipxe/src'
  [BUILD] bin-arm64-efi/__umoddi3.o
  [BUILD] bin-arm64-efi/__udivdi3.o
  [BUILD] bin-arm64-efi/__moddi3.o
  [BUILD] bin-arm64-efi/__divmoddi4.o
  [BUILD] bin-arm64-efi/icc.o
  [BUILD] bin-arm64-efi/__divdi3.o
  [BUILD] bin-arm64-efi/implicit.o
  [BUILD] bin-arm64-efi/__udivmoddi4.o
  [BUILD] bin-arm64-efi/pc_kbd.o
In file included from include/ipxe/iomap.h:16,
                 from include/ipxe/io.h:23,
                 from core/pc_kbd.c:13:
./config/ioapi.h:19:10: fatal error: config/local/ioapi.h: No such file or directory
   19 | #include <config/local/ioapi.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~

**Comment by NiKiZe** [comment]
ipxe.efi is not a working target on arm, please build for your specific NIC or snp.efi

**Reply by mcb30** [reply]
> ipxe.efi is not a working target on arm, please build for your specific NIC or snp.efi

Actually no longer true as of https://github.com/ipxe/ipxe/commit/c2226b3d1 - we now have dummy `outb()` etc on ARM specifically to make it possible to build `ipxe.efi`.

**Comment by mcb30** [comment]
I'm not sure why your build is failing.  You can definitely simplify your build command to
```
make CROSS=aarch64-linux-gnu- bin-arm64-efi/ipxe.efi EMBED=boot.ipxe
```
(i.e. omit the `ARCH=arm64` and use `CROSS=` instead of `CROSS_COMPILE=`).

I don't know if this will help, but it's generally a good idea to simplify as much as possible before starting to debug.

**Comment by NiKiZe** [comment]
Sorry for not reading properly this sounds like the issue #1265 tried to fix, but couldn't reproduce. (And we have also seen before)

If you are still having issues with missing config/local files, try a clean checkout. If that doesn't work tell us more about your environment.

**Comment by jamaya77** [comment, marked as answer]
Thank you, Christian/Michael for your time.

I ran the suggested command and generated this message at the end, possibly because the "boot.ipxe" file doesn't exist:
make: *** No rule to make target 'boot.ipxe', needed by 'bin-arm64-efi/embedded.o'.  Stop.

However, despite the message, the "ipxe.efi" file was created in the /ipxe/src/bin-arm64-efi directory and copied the file:
cp -r bin-arm64-efi/ipxe.efi /tftpboot/ipxe_arm.efi

Configured the dnsmasq.conf file:

#### # if request comes from firmware, load iPXE via TFTP
pxe-service=tag:!ENH,X86-64_EFI, iPXE Boot (UEFI), ipxe.efi
pxe-service=tag:!ENH,X86PC, iPXE Boot (BIOS), undionly.kpxe
pxe-service=tag:!ENH,ARM64_EFI, iPXE Boot (ARM64 UEFI), ipxe_arm.efi

#### # if request comes from iPXE, direct it to boot from boot.ipxe
#### # dhcp-boot=tag:ENH,http://192.168.1.21/boot.ipxe
pxe-service=tag:ENH,X86-64_EFI, UEFI: ipxe.efi - iPXE, http://192.168.1.21/menu/bootuefi.ipxe
pxe-service=tag:ENH,X86PC, PC-DOS: undionly.kpxe - iPXE, http://192.168.1.21/menu/bootbios.ipxe
pxe-service=tag:ENH,ARM64_EFI, UEFI: ipxe_arm.efi - iPXE, http://192.168.1.21/menu/bootuefi.ipxe

To be honest, I don't know if this configuration is correct, but I'm just replicating the original configuration. Now, when booting the arm64 device, ipxe is responding, but generating a "autoexec.ipxe....not found" message. Suggestions? Thank you 


![Screenshot from 2024-09-16 14-19-40](https://github.com/user-attachments/assets/f9f846f1-a6e4-420e-86f8-c925c1289277)

**Reply by NiKiZe** [reply]
You are telling it to embed `boot.ipxe`, since you have boot configuration in DHCP you probable don't want or need it, but it depends on your setup, but you should adjust your make command accordingly.

Other than this, anything that isn't working? (You can ignore autoexec message, it's information, not an error)

**Reply by jamaya77** [reply]
Thank you, Christian. I'm not sure what's missing or if it's bios configuration on the laptop itself. The issue I'm having now is that the ipxe boot menu is not loading on the arm64 laptop. My configuration currently is separate boot menu for BIOS and UEFI based devices as defined in my last post. The boot menu also offers to boot directly to SCCM via chainloading depending if its a BIOS or UEFI device and it's working well. Will probably have to the same for arm64.

chain tftp://${wdsserver}/smsboot\x64\wdsnbp.com

chain tftp://${wdsserver}/smsboot\x64\wdsmgfw.efi

When I look in the /var/log/dnsmasq.log, I see a BIOS device (192.168.1.33) that successfully received the boot menu and the the arm64 device (192.168.1.17) that did not. Sorry if I provided too much info.

Sep 17 11:22:18 dnsmasq-dhcp[2945]: PXE(ens18) 08:00:47:3e:aa:9f proxy
Sep 17 11:22:18 dnsmasq-dhcp[2945]: PXE(ens18) 08:00:47:3e:aa:9f proxy
Sep 17 11:22:22 dnsmasq-dhcp[2945]: PXE(ens18) 192.168.1.33 08:00:47:3e:aa:9f undionly.kpxe
Sep 17 11:22:23 dnsmasq-tftp[2945]: error 0 TFTP Aborted received from 192.168.1.33
Sep 17 11:22:23 dnsmasq-tftp[2945]: failed sending /tftpboot/undionly.kpxe to 192.168.1.33
Sep 17 11:22:23 dnsmasq-tftp[2945]: sent /tftpboot/undionly.kpxe to 192.168.1.33
Sep 17 11:22:25 dnsmasq-dhcp[2945]: PXE(ens18) 08:00:47:3e:aa:9f proxy
Sep 17 11:22:25 dnsmasq-dhcp[2945]: PXE(ens18) 192.168.1.33 08:00:47:3e:aa:9f http://192.168.1.21/menu/bootbios.ipxe


Sep 17 11:36:17 dnsmasq-dhcp[2945]: PXE(ens18) d8:ec:4e:85:cf:8f proxy
Sep 17 11:36:20 dnsmasq-dhcp[2945]: PXE(ens18) d8:ec:4e:85:cf:8f proxy
Sep 17 11:36:21 dnsmasq-tftp[2945]: error 8 User aborted the transfer received from 192.168.1.17
Sep 17 11:36:21 dnsmasq-tftp[2945]: failed sending /tftpboot/ipxe_arm.efi to 192.168.1.17
Sep 17 11:36:23 dnsmasq-tftp[2945]: sent /tftpboot/ipxe_arm.efi to 192.168.1.17
Sep 17 11:36:23 dnsmasq-tftp[2945]: file /tftpboot/autoexec.ipxe not found




![bootmenu](https://github.com/user-attachments/assets/342fa603-9ee2-4818-a247-be356664c013)

**Comment by jamaya77** [comment]
Finally got it working! Had to go a little more basic in creating the arm64 "ipxe.efi" file and referred to my [ticket](https://github.com/ipxe/ipxe/discussions/293) from three years ago when I first started using ipxe.

1) Download ipxe
2) Edited "general.h" file, /ipxe/src/config/general.h
3) Edited "console.h" file, /ipxe/src/config/console.h
4) Compiled a new ipxe.efi file:
make CROSS=aarch64-linux-gnu- bin-arm64-efi/ipxe.efi

Thank you, Christian and Michael, for your time and support!

**Reply by matthewei** [reply]
Hi @jamaya77 could you share your method to build arm64 ipxe?

**Comment by jamaya77** [comment]
Hi @matthewei 

Sure, this is a copy/paste from the history on my linux server.
The last three "make" commands compiles the files for:
1) Bios or legacy devices (undionly.kpxe)
2) UEFI devices (ipxe.efi)
3) ARM64 devices (ipxe.efi) which I later rename to "ipxe_arm.efi"

Should be able see these directories after downloading ipxe in /ipxe/src folder:

[root@ipxe src]# ls -d bin* config*
bin   bin-arm64-efi   bin-x86_64-efi   config



git clone https://github.com/ipxe/ipxe.git
cd ipxe/src
vim config/general.h 
vim config/console.h 
make bin/undionly.kpxe
make bin/undionly.kpxe bin-x86_64-efi/ipxe.efi
make CROSS=aarch64-linux-gnu- bin-arm64-efi/ipxe.efi

**Reply by NiKiZe** [reply]
The bin directories does not exist until after running the make command for the given directory. So nothing wrong if they are missing on a git checkout.

For configs I recommend using the files in local, Ur using named configs https://ipxe.org/appnote/named_config

**Reply by matthewei** [reply]
Thanks all , Could you share your dnsmasq's configurations? By the way, Does  arm64 need bin/undionly.kpxe?

**Reply by matthewei** [reply]
@jamaya77 Looking forward to hear from you!
